mock/py/mockbuild/plugins/overlayfs.py

```diff
@@ -0,0 +1,815 @@
+        lowerList = self.createLayerList(lowerTopLayerId)
+
+        workDir = self.getWorkDir()
+        if os.path.exists(workDir):
+            shutil.rmtree(workDir)
+        os.mkdir(workDir)
+
+        # make sure kernel has required module loaded
+        modprobeCmds = []
+        modprobeCmds.append("modprobe")
+        modprobeCmds.append("overlay")
+        subprocess.check_call(modprobeCmds)
+
+        mountCmds = []
+        mountCmds.append("mount")
+        mountCmds.append("-t")
+        mountCmds.append("overlay")
+        mountCmds.append("overlay")
+
+        optionsArg="-olowerdir="
+
+        firstLower=True
+        for lowerId in lowerList:
+            if not firstLower:
+                optionsArg += ":"
+            firstLower = False
+            optionsArg += self.getLayerFsDir(lowerId)
+
+        optionsArg += ",upperdir=" + self.getLayerFsDir(upperLayerId)
+        optionsArg += ",workdir=" + workDir
+
+        mountCmds.append(optionsArg)
+        mountCmds.append(self.getRootDir())
+        subprocess.check_call(mountCmds)
+
+        self.recordRootMounted(True)
+
+
+    # unmount root
+    def unmountRoot(self):
+        if self.isRootMounted():
+            umountCmds = []
+            umountCmds.append("umount")
+            umountCmds.append(self.getRootDir())
+            subprocess.check_call(umountCmds)
+
+            self.recordRootMounted(False)
+            workDir = self.getWorkDir()
+            if os.path.exists(workDir):
+                shutil.rmtree(workDir)
+
+
+    def recordRootMounted(self, mounted):
+        rootMountFlagFile = self.rootMountFlagFile()
+        if mounted:
+            self.writeFile(rootMountFlagFile, "")
+        else:
+            os.remove(rootMountFlagFile)
+
+
+    def isRootMounted(self):
+        rootMountFlagFile = self.rootMountFlagFile()
+        isRootMounted = os.path.exists(rootMountFlagFile)
+        return isRootMounted
+
+    # lock on snapshot operations ( used to prevent concurent modification of
+    # refs/layers by mock )
+    def snapshotLock(self):
+        snapshotLockFile = self.getSnapshotLockFile()
+        try:
+            os.mkdir(snapshotLockFile)
+        except OSError:
+            raise Exception("Failed to obtain snapshot lock !")
+
+    def snapshotUnlock(self):
+        snapshotLockFile = self.getSnapshotLockFile()
+        if os.path.exists(snapshotLockFile):
+            os.rmdir(snapshotLockFile)
+
+    # lock on mount operations
+    def mountLock(self):
+        mountLockFile = self.getMountLockFile()
+        try:
+            os.mkdir(mountLockFile)
+        except OSError:
+            raise Exception("Failed to obtain mount lock !")
+
+    def mountUnlock(self):
+        mountLockFile = self.getMountLockFile()
+        os.rmdir(mountLockFile)
+
+    def debugPrint(self, message):
+        if self.debug:
+            sys.stderr.write("DEBUG: Overalyfs pluin: " + message + "\n")
```

Comment:
Instead of direct print to stdderr, it is better to use set up logger (which can point somewhere else)
`self.buildroot.root_log.debug("foo")`